llmfoundry/data/packing.py

```diff
@@ -260,14 +260,142 @@ def pad_tensor(tensor: torch.Tensor, pad_value: int):
+        tokenizer (PreTrainedTokenizerBase): The tokenizer for profiling.
+        device_batch_size (int): The size of the batches (number of examples) per device.
+
+    Returns:
+        A packing ratio that minimizes padding while maintaining zero waste.
+    """
+    min_ratio = 1
+    max_ratio = 20
+    num_packing_ratios = 10
+    profiling_results = profile_packing(dataloader_cfg, tokenizer, min_ratio,
+                                        max_ratio, num_packing_ratios,
+                                        device_batch_size)
+
+    # Obtain the maximum packing_ratio/minimum padding that has no waste.
+    i = 0
+    waste = 0
+    packing_ratio = 1
+    while i < len(profiling_results) and waste == 0:
+        packing_ratio, _, waste = profiling_results[i]
+        i += 1
+    return packing_ratio
+
+
+def profile_packing(dataloader_cfg: DictConfig,
+                    tokenizer: PreTrainedTokenizerBase, min_ratio: float,
+                    max_ratio: float, num_packing_ratios: int,
+                    device_batch_size: int):
+    """Profile packing.
+
+    Args:
+        dataloader_cfg (DictConfig): The dataloader configuration for profiling.
+        tokenizer (PreTrainedTokenizerBase): The tokenizer for profiling.
+        min_ratio (float): Smallest packing_ratio to test. Must be >=1.
+        max_ratio (float): Largest packing_ratio to test. Must be larger than `min_ratio`.
+        num_packing_ratios (int): Number of packing_ratio values (spaced between `min_ratio` and `max_ratio`) to try.
+        device_batch_size (int): The size of the batches (number of examples) per device.
+
+    Returns:
+        A list of tuples of packing ratio, padding, and waste.
+    """
+    import copy
+
+    from llmfoundry import (build_finetuning_dataloader,
+                            build_text_denoising_dataloader)
+    from llmfoundry.data import build_text_dataloader
+
+    # Turn off packing for the dataloader (we want raw, pre-packed examples)
+    dataloader_cfg = copy.deepcopy(dataloader_cfg)
+    dataloader_cfg.dataset.packing_ratio = None
+    dataloader_cfg.dataset.max_leftovers_to_keep = None
+
+    # Determine the packing_ratio values we'll try
+    packing_ratios, raw_batch_sizes = [], []
+    for packing_ratio in np.linspace(min_ratio,
+                                     max_ratio,
+                                     num_packing_ratios,
+                                     endpoint=True):
+        packing_ratio = np.round(10 * packing_ratio) / 10
+        raw_batch_size = int(packing_ratio * device_batch_size)
+        if raw_batch_size not in raw_batch_sizes:
+            packing_ratios.append(packing_ratio)
+            raw_batch_sizes.append(raw_batch_size)
+
+    examples_to_profile = max(
+        raw_batch_sizes
+    ) * 100  # TODO: This should be min'd with dataset size, but how do we get dataset size before we construct it? Might need a refactor
```

Comment:
Error that we want to avoid:
```
ValueError: Your dataset (name=HuggingFaceH4/databricks_dolly_15k, split=train) has 14989 samples, but your minimum batch size is 20000 because you are running on 1 gpus and your per 
device batch size is 20000. Please increase the number of samples in your dataset to at least 20000.
```

repo:
```
python llmfoundry/data/packing.py --yaml-path scripts/train/yamls/finetune/7b_dolly_sft.yaml --min 50 --max 50 --num-devices 16
```